text/0000-const-type-inference.md

```diff
@@ -0,0 +1,114 @@
+- Feature Name: const_type_inference
+- Start Date: 2023-12-21
+- RFC PR: [rust-lang/rfcs#3546](https://github.com/rust-lang/rfcs/pull/3546)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Allow type inference for `const` or `static` when the type of the initial value is known.
+
+# Motivation
+[motivation]: #motivation
+
+Rust currently requires explicit type annotations for `const` and `static` items.
+
+
+In simple cases, explicitly writing out
+the type of the const seems trivial. However, this isn't always the case:
+
+- Sometimes the constant's value is complex, making the explicit type overly verbose.
+- In some cases, the type may be unnameable.
+- When creating macros, the precise type might not be known to the macro author.
+- Code generators may not have enough information to easily determine the type.
+
+This change aims to make Rust code more concise and maintainable, especially in scenarios where the types of
+const items are complicated or not easily expressible.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+You may declare constants and static variables without specifying their types when the type can be inferred
+from the initial value. For example:
+
+```rs
+const PI = 3.1415; // inferred as f64
```

Comment:
I would personally prefer that the actual type of numeric or float types be known rather than picking arbitrary defaults (e.g. `i32` or `f64)`. I'm not too keen on it in a local context either tbh but there it's mitigated by the compiler being able to infer the real type from the surrounding code most of the time.
```rust
const PI = 3.1415; // error
const PI = 3.1415_f64; // inferred as f64
```

Reply:
I would prefer more consistent behavior with let bindings.  So let's make a simple vote here:
🎉:  `const PI = 3.1415; // error`
🚀: `const PI = 3.1415; // inferred as f64`

Reply:
It's not a big issue for me either way it's just that a few times I've only later realised a type has been unexpectedly made  `i32`. But it's easily fixed and more of an annoyance than a problem per se. It doesn't help that an `i32` is very rarely what I want.

Reply:
If it follows the normal rust inference default types then it's at least no *worse* than what `let` does.

Reply:
Well it's worse in the sense that let can use future code to infer the type, so mostly this is a non issue there unless there's a lot of generics involved.

Reply:
Honestly I would most like if the const was of `{float}` type but people aren't ready for that conversation maybe